### PR TITLE
[Merged by Bors] - tortoise: stay in verifying mode while recovering from disk

### LINKED ISSUE #5246: count ballots in the verifying mode on the tortoise recovery code path
it will significantly reduce cpu usage on restart.
it requires triage, as it is not clear why clear why it counts ballots in full mode.

### PR DESCRIPTION
**Pull request by dshulyak**
closes https://github.com/spacemeshos/go-spacemesh/issues/5246

verifying mode is a faster way to count ballots, and it is also has fewer known security issues.

during debug i observed the following problems, which are mostly fixed here:
- certificates are not kept forever and hence tortoise cannot form an opinion on the layer without them. to correctly handle that we update last layer and opinion is loaded the database.
- when data is recovered from disk we need to start counting from the first layer in the epoch, in order to compute reference epoch height. this variable is used later in verifyLayers call 
- after layers were recovered we compute matching opinions and evict unnecessary events. this eviction was too agressive and should be limited to layers that were verified by tortoise. this is to avoid the same race as in https://github.com/spacemeshos/go-spacemesh/pull/5523

additionally i switched tortoise to use IterateAtxs directly on the database without using lru cache, as it is faster and intend to work on dropping that in other components.

### CODE REVIEW

File: sql/blocks/blocks.go [hunk 1]
```diff
@@ -71,6 +71,25 @@ func Get(db sql.Executor, id types.BlockID) (rst *types.Block, err error) {
 	return rst, err
 }
 
+func LastValid(db sql.Executor) (types.LayerID, error) {
+	var lid types.LayerID
+	// it doesn't use max(layer) in order to get rows == 0 when there are not layers
```
Comment (poszu) [inline sql/blocks/blocks.go]: I don't understand the subtle difference here, could you explain?

Btw a typo:
```suggestion
	// it doesn't use max(layer) in order to get rows == 0 when there are no layers
```

File: sql/blocks/blocks_test.go [hunk 2]
```diff
@@ -217,3 +217,48 @@ func TestGetLayer(t *testing.T) {
...
+			)
+			require.NoError(t, Add(db, block))
+		}
+		for i := range ids[:cutoff] {
+			require.NoError(t, SetValid(db, ids[i]))
+		}
+		last, err := LastValid(db)
+		require.NoError(t, err)
+		require.Equal(t, layers[cutoff-1], last)
+	})
```
Comment (poszu) [inline sql/blocks/blocks_test.go]: I find the following form of this test plain and easier to understand (I don't need to figure out what is cutoff for instance):
```suggestion
	t.Run("last is invalid", func(t *testing.T) {
		db := sql.InMemory()

		blocks := map[types.BlockID]struct {
			invalid bool
			lid     types.LayerID
		}{
			{1}: {lid: 11},
			{2}: {lid: 22},
			{3}: {invalid: true, lid: 33},
		}

		for bid, layer := range blocks {
			block := types.NewExistingBlock(
				bid,
				types.InnerBlock{LayerIndex: layer.lid},
			)
			require.NoError(t, Add(db, block))
			if !layer.invalid {
				require.NoError(t, SetValid(db, bid))
			}
		}

		last, err := LastValid(db)
		require.NoError(t, err)
		require.Equal(t, 22, last)
	})
```

Comment (poszu) [inline sql/blocks/blocks_test.go]: I find the following form of this test, where related data is kept together,  plain and easier to understand (for instance, I don't need to figure out what cutoff is for):
```suggestion
	t.Run("last is invalid", func(t *testing.T) {
		db := sql.InMemory()

		blocks := map[types.BlockID]struct {
			invalid bool
			lid     types.LayerID
		}{
			{1}: {lid: 11},
			{2}: {lid: 22},
			{3}: {invalid: true, lid: 33},
		}

		for bid, layer := range blocks {
			block := types.NewExistingBlock(
				bid,
				types.InnerBlock{LayerIndex: layer.lid},
			)
			require.NoError(t, Add(db, block))
			if !layer.invalid {
				require.NoError(t, SetValid(db, bid))
			}
		}

		last, err := LastValid(db)
		require.NoError(t, err)
		require.Equal(t, 22, last)
	})
```

File: tortoise/algorithm.go [hunk 3]
```diff
@@ -140,7 +140,7 @@ func (t *Tortoise) RecoverFrom(lid types.LayerID, opinion, prev types.Hash32) {
 	t.trtl.evicted = lid - 1
 	t.trtl.pending = lid
 	t.trtl.verified = lid
-	t.trtl.processed = lid
+	t.trtl.processed = lid - 1 // -1 so that iteration in tallyVotes starts from the target layer
```
Comment (poszu) [inline tortoise/algorithm.go:149]: Could `RecoverFrom(lid)` be called with `lid=0`?

File: tortoise/tracer.go [hunk 4]
```diff
@@ -445,6 +441,46 @@ func (m *MalfeasanceTrace) Run(r *traceRunner) error {
 	return nil
 }
 
+type headerWithValidity struct {
+	Headers  types.BlockHeader `json:"header"`
+	Validity bool              `json:"validity"`
+}
```
Comment (poszu) [inline tortoise/tracer.go]: `{"validity": true`} looks weird. How about
```suggestion
type headerWithValidity struct {
	Header types.BlockHeader `json:"header"`
	Valid  bool              `json:"valid"`
}
```

File: tortoise/recover.go [hunk 5]
```diff
@@ -20,7 +19,7 @@ import (
 // Recover tortoise state from database.
 func Recover(
 	ctx context.Context,
-	db *datastore.CachedDB,
+	db *sql.Database,
```
Comment (poszu) [inline tortoise/recover.go]: How about using `sql.Executor` where possible?
```suggestion
	db sql.Executor
```

File: tortoise/recover.go [hunk 6]
```diff
@@ -114,19 +113,33 @@ func Recover(
...
+		}
+	}
 	return trtl, nil
 }
 
-func recoverEpoch(epoch types.EpochID, trtl *Tortoise, db *datastore.CachedDB) error {
-	if err := db.IterateEpochATXHeaders(epoch, func(header *types.ActivationTxHeader) error {
-		trtl.OnAtx(header.ToData())
-		return nil
+func recoverEpoch(target types.EpochID, trtl *Tortoise, db *sql.Database) error {
```
Comment (poszu) [inline tortoise/recover.go]: Ditto
```suggestion
func recoverEpoch(target types.EpochID, trtl *Tortoise, db sql.Executor error {
```

File: tortoise/replay/replay_test.go [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+	require.NoError(t, err)
+	clock, err := timesync.NewClock(
+		timesync.WithLayerDuration(cfg.LayerDuration),
+		timesync.WithTickInterval(1*time.Second),
+		timesync.WithGenesisTime(genesis),
+		timesync.WithLogger(log.NewNop()),
+	)
+	require.NoError(t, err)
+
+	db, err := sql.Open("file:" + *dbpath)
```
Comment (poszu) [inline tortoise/replay/replay_test.go]: Would it make sense to open the DB in read-only mode?

File: tortoise/tortoise.go [hunk 8]
```diff
@@ -506,30 +518,38 @@ func (t *turtle) runFull() (verified, changed types.LayerID) {
...
+		zap.Uint64("height", einfo.height),
+	)
 }
 
 func (t *turtle) onBlock(header types.BlockHeader, data, valid bool) {
 	if header.LayerID <= t.evicted {
 		return
 	}
+
+	t.logger.Debug("on block", zap.Inline(&header), zap.Bool("data", data), zap.Bool("valid", valid))
```
Comment (poszu) [inline tortoise/tortoise.go:543]: I find it confusing that this is called "data":
```go
type blockInfo struct {
    ...
    data bool // set to true if block is available locally
}
```
Nobody without good knowledge of the code will understand this log message and the meaning of `data` in general.
```suggestion
	t.logger.Debug("on block", zap.Inline(&header), zap.Bool("available", available), zap.Bool("valid", valid))
```

File: sql/blocks/blocks.go (hunk 1 above)
Comment (dshulyak) [reply sql/blocks/blocks.go]: max(layer) aggregation will always return rows = 1, even if nothing exists. i will add to the comment

File: tortoise/algorithm.go (hunk 3 above)
Comment (dshulyak) [reply tortoise/algorithm.go:149]: it can't be earlier than genesis. i will add assertion with panic

File: tortoise/tortoise.go (hunk 8 above)
Comment (dshulyak) [reply tortoise/tortoise.go:543]: i will keep it as data, as i refer to it as such accross the code. it is meant to be a debug log, i don't expect it to be clear to anyone without looking at code

File: tortoise/replay/replay_test.go (hunk 7 above)
Comment (dshulyak) [reply tortoise/replay/replay_test.go]: yes, added mode=ro

**Review by poszu — APPROVED** [review]

**Comment by dshulyak** [comment]
bors merge

**Comment by dshulyak** [comment]
bors merge

**Pull request closed without merging**